admin/cluster_setup.py

```diff
@@ -0,0 +1,286 @@
+
+        self['purpose'] = unicode(self['purpose'])
+        if any(x not in string.ascii_letters + string.digits + '-'
+               for x in self['purpose']):
+            raise UsageError(
+                "Purpose may have only alphanumeric symbols and dash. " +
+                "Found {!r}".format('purpose')
+            )
+        # This is run last as it creates the actual "runner" object
+        # based on the provided parameters.
+        super(RunOptions, self).postOptions()
+
+    def _make_cluster_identity(self, dataset_backend):
+        purpose = self['purpose']
+        return ClusterIdentity(
+            purpose=purpose,
+            prefix=purpose,
+            name='{}-cluster'.format(purpose).encode("ascii"),
+        )
+
+
+@inlineCallbacks
+def main(reactor, args, base_path, top_level):
+    """
+    :param reactor: Reactor to use.
+    :param list args: The arguments passed to the script.
+    :param FilePath base_path: The executable being run.
+    :param FilePath top_level: The top-level of the Flocker repository.
+    """
+    options = RunOptions(top_level=top_level)
+
+    add_destination(eliot_output)
+    try:
+        options.parseOptions(args)
+    except UsageError as e:
+        sys.stderr.write("%s: %s\n" % (base_path.basename(), e))
+        raise SystemExit(1)
+
+    runner = options.runner
+
+    from flocker.common.script import eliot_logging_service
+    log_writer = eliot_logging_service(
+        destination=FileDestination(
+            file=open("%s.log" % (base_path.basename(),), "a")
+        ),
+        reactor=reactor,
+        capture_stdout=False)
+    log_writer.startService()
+    reactor.addSystemEventTrigger(
+        'before', 'shutdown', log_writer.stopService)
+
+    cluster = None
+    results = []
+    try:
+        yield runner.ensure_keys(reactor)
+        cluster = yield runner.start_cluster(reactor)
+        if options['distribution'] in ('centos-7',):
+            remote_logs_file = open("remote_logs.log", "a")
+            for node in cluster.all_nodes:
+                results.append(capture_journal(reactor,
+                                               node.address,
+                                               remote_logs_file)
+                               )
+        elif options['distribution'] in ('ubuntu-14.04', 'ubuntu-15.10'):
+            remote_logs_file = open("remote_logs.log", "a")
+            for node in cluster.all_nodes:
+                results.append(capture_upstart(reactor,
+                                               node.address,
+                                               remote_logs_file)
+                               )
+        gather_deferreds(results)
+
+        if options['apps-per-node'] > 0:
+            config = _build_config(cluster, options['template'],
+                                   options['apps-per-node'])
+            yield _configure(reactor, cluster, config)
+
+        result = 0
+
+    except BaseException:
```

Comment:
Yes, it's explained a little bit here bb131ed424a5c7d. In `acceptance.py` we have a naked `except:` in that place.
Specifically, I've seen a situation where `GeneratorExit` was injected into `main`.  And we want to set `result` in those cases, so that we could clean up the cluster.